source/module_base/test/complexarray_test.cpp

```diff
@@ -45,125 +50,416 @@
+    a.randomize();
+
+    bool allequal = true;
+    for (int i=1;i<5;++i)
+    {
+        allequal = allequal && (a.ptr[i] == a.ptr[0]);
+    }
+    EXPECT_FALSE(allequal);
+
+    for (int i=0;i<a.getSize();++i)
+    {
+        EXPECT_TRUE((a.ptr[i].real() >= -0.5) && (a.ptr[i].imag() >= -0.5));
+        EXPECT_TRUE((a.ptr[i].real() < 0.5) && (a.ptr[i].imag() < 0.5));
+    }
+
+}
+
+TEST(ComplexArray,getBoundSize)
+{
+    ModuleBase::ComplexArray a(1,2,3,4);
+    EXPECT_EQ(a.getSize(),24);
+    EXPECT_EQ(a.getBound1(),1);
+    EXPECT_EQ(a.getBound2(),2);
+    EXPECT_EQ(a.getBound3(),3);
+    EXPECT_EQ(a.getBound4(),4);
+}
+
+
+TEST_F(ComplexArray_test,create)
+{
+    a2.create(2,3,4,5);
+    EXPECT_EQ(a2.getSize(),120);
+}
+
+TEST_F(ComplexArray_test,operator_double_multiply)
+{
+    a2 = com1;
+    c2 = 2.0 * a2 ;
+    for (int i = 0;i<a2.getSize();++i)
+    {
+        EXPECT_COMPLEX_EQUAL(c2.ptr[i],2.0 * com1);
+    }  
+}
+
+TEST_F(ComplexArray_test,operator_complex_multiply)
+{
+    a2 = com1;
+    c2 = com2 * a2 ;
+    for (int i = 0;i<a2.getSize();++i)
+    {
+        EXPECT_COMPLEX_EQUAL(c2.ptr[i],com2 * com1);
+    }  
+}
+
+TEST(ComplexArray,abs2)
+{
+    ModuleBase::ComplexArray a(2,2,1,1);
+    a(0,0,0,0) = std::complex<double>{0.0,0.0};
+    a(1,0,0,0) = std::complex<double>{1.0,1.0};
+    a(0,1,0,0) = std::complex<double>{2.0,2.0};
+    a(1,1,0,0) = std::complex<double>{3.0,3.0};
+    EXPECT_DOUBLE_EQ(abs2(a),28.0);
+}
+
+TEST(ComplexArray,dot)
+{
+    ModuleBase::ComplexArray a(2,2,1,1);
+    a(0,0,0,0) = std::complex<double>{0.0,1.0};
+    a(1,0,0,0) = std::complex<double>{1.0,0.0};
+    a(0,1,0,0) = std::complex<double>{2.0,3.0};
+    a(1,1,0,0) = std::complex<double>{3.0,2.0};
+
+    ModuleBase::ComplexArray b(2,2,1,1);
+    b(0,0,0,0) = std::complex<double>{1.0,0.0};
+    b(1,0,0,0) = std::complex<double>{2.0,-1.0};
+    b(0,1,0,0) = std::complex<double>{3.0,-2.0};
+    b(1,1,0,0) = std::complex<double>{4.0,-3.0};
+    std::complex<double> expectab {8.0,-32.0};
+    std::complex<double> expectba {8.0,32.0};
+    EXPECT_COMPLEX_EQUAL(dot(a,b),expectab);
+    EXPECT_COMPLEX_EQUAL(dot(b,a),expectba);
+}
+
+TEST_F(ComplexArray_test,scale_accumulate_double)
+{
+    a2 = com1;
+    b2 = com2;
+    scale_accumulate(0.3,a2,b2);
+    for (int i = 0;i<b2.getSize();++i)
+    {
+        EXPECT_COMPLEX_EQUAL(b2.ptr[i],0.3*com1+com2);
+    }  
+}
+
+TEST_F(ComplexArray_test,scale_accumulate_complex)
+{
+    a2 = com1;
+    b2 = com2;
+    scale_accumulate(com3,a2,b2);
+    for (int i = 0;i<b2.getSize();++i)
+    {
+        EXPECT_COMPLEX_EQUAL(b2.ptr[i],com3*com1+com2);
+    }  
+}
+
+TEST_F(ComplexArray_test,scaled_sum)
+{
+    a2 = com1;
+    b2 = com2;
+    scaled_sum(0.3,a2,0.4,b2,c2);
+    for (int i = 0;i<c2.getSize();++i)
+    {
+        EXPECT_COMPLEX_EQUAL(c2.ptr[i],0.3*com1 + 0.4*com2);
+    }  
+}
+
+TEST_F(ComplexArray_test,point_mult)
+{
+    a2 = com1;
+    b2 = com2;
+    point_mult(a2,b2,c2);
```

Comment:
The name `point_mult` is really weird here - being something sounds like "dot product" but actually no. Any suggestions?
> .* is Element-wise multiplication follow rules for array operations
 Also called: Hadamard Product, Schur Product and broadcast

Reply:
Agree, I think hadamard_product is good.

Reply:
I just searched it and found that there has another name that is pointwise multiplication. In that light, abbreviated as point_mult is not unacceptable.
https://math.stackexchange.com/questions/20412/element-wise-or-pointwise-operations-notation